backend/ttrs/recommend2.py

```diff
@@ -0,0 +1,459 @@
+    available = []
+    for i, lecture in enumerate(whole_lectures):
+        available.append(contains(info['blocks'], lecture.time_slot_set))
+
+    # Init lectures
+    lectures = []
+    for i, lecture in enumerate(whole_lectures):
+        if not available[i]:
+            continue
+        next_lectures = []
+        for j in whole_compatible[i]:
+            if available[j]:
+                next_lectures.append(whole_lectures[j])
+        lecture.init(next_lectures, info)
+        lectures.append(lecture)
+
+    return lectures, info
+
+
+def save(year, semester):
+    """
+    Save list of SimpleLecture object to external resource (using pickle).
+    """
+    lectures = [SimpleLecture(lecture) for lecture in Lecture.objects.filter(year=year, semester=semester) if lecture.time_slots.count() > 0]
+    with open('lectures{}{}.pickle'.format(year, semester), 'wb') as f:
+        pickle.dump(lectures, f)
+    distinct = []
+    for i, lecture1 in enumerate(lectures):
+        indices = []
+        for j in range(i + 1, len(lectures)):
+            lecture2 = lectures[j]
+            if lecture1.course != lecture2.course and not intersect(lecture1.time_slot_set, lecture2.time_slot_set):
+                indices.append(j)
+        distinct.append(indices)
+        print('{}/{}'.format(i, len(lectures)), len(indices), len(indices)*100//len(lectures))
+    with open('compatible{}{}.pickle'.format(year, semester), 'wb') as f:
+        pickle.dump(distinct, f)
+
+
+def load(year, semester):
+    """
+    Load list of SimpleLecture object and their compatible lecture list
+    from external resource  (using pickle).
+    """
+    try:
+        with open('lectures{}{}.pickle'.format(year, semester), 'rb') as f:
+            whole_lectures = pickle.load(f)
+        with open('compatible{}{}.pickle'.format(year, semester), 'rb') as f:
+            whole_compatible = pickle.load(f)
+    except Exception as e:
+        print(e)
+        return None, None
+    return whole_lectures, whole_compatible
+
+
+def rank_lecture_set(lectures: List[SimpleLecture], info: dict):
+    """
+    Using A* algorithm, rank lecture sets and return the list.
+    :param lectures: Whole available lecture list
+    :param info: information about the recommendation
+    :return: Ranked lecture set list
+    """
+    contexts = []
+    base = Context()
+    for lecture in lectures:
+        heapq.heappush(contexts, Context(base, lecture, info))
+    ranks = []
+    max_score = -1000000000
+    for count in range(500):
+        if not contexts:
+            break
+        context = heapq.heappop(contexts)
+        final_score = context.get_final_score(info)
+        max_score = max(max_score, final_score)
+        ranks.append((context, final_score))
+
+        # print(count, final_score, context)
+        for lecture in context.lectures[-1].next_lectures:
+            if context.compatible(lecture) and context.credit + lecture.credit <= info['expected_credit']:
```

Comment:
This isn't a major issue, but I thought expected credit not as upper bound but as _expectation_. By that I mean, 
if my expected credit is 16, I can accept credits 15~17 or somewhere around.

Reply:
I think so...
But, I just did as you have done.
I'll add 2 credit padding, okay?